src/game/logic/object/experiencetracker.cpp

```diff
@@ -0,0 +1,182 @@
+    if (killer->Get_Relationship(m_parent) == Relationship::ALLIES) {
+        return 0;
+    }
+
+    return m_parent->Get_Template()->Get_Experience_Value(m_currentLevel);
+}
+
+// zh: 0x0061B7D0 wb: 0x008A3637
+bool ExperienceTracker::Is_Trainable() const
+{
+    return m_parent->Get_Template()->Is_Trainable();
+}
+
+// zh: 0x0061B7E0 wb: 0x008A3654
+bool ExperienceTracker::Is_Accepting_Experience_Points() const
+{
+    return Is_Trainable() || m_experienceSink != ObjectID::OBJECT_UNK;
+}
+
+// zh: 0x0061B820 wb: 0x008A36A4
+void ExperienceTracker::Set_Min_Veterency_Level(VeterancyLevel new_level)
+{
+    if (m_currentLevel >= new_level) {
+        return;
+    }
+
+    captainslog_dbgassert(m_parent != nullptr, "Parent must not be nullptr!");
+
+    const auto old_level = m_currentLevel;
+    m_currentLevel = new_level;
+
+    m_currentExperience = m_parent->Get_Template()->Get_Experience_Required(m_currentLevel);
+
+    m_parent->On_Veterancy_Level_Changed(old_level, m_currentLevel, true);
+}
+
+// zh: 0x0061B860 wb: 0x008A370D
+void ExperienceTracker::Set_Veterency_Level(VeterancyLevel new_level, bool unk)
+{
+    if (m_currentLevel == new_level) {
+        return;
+    }
+
+    captainslog_dbgassert(m_parent != nullptr, "Parent must not be nullptr!");
+
+    const auto old_level = m_currentLevel;
+    m_currentLevel = new_level;
+
+    m_currentExperience = m_parent->Get_Template()->Get_Experience_Required(m_currentLevel);
+
+    m_parent->On_Veterancy_Level_Changed(old_level, m_currentLevel, true);
+}
+
+// zh: 0x0061B8A0 wb: 0x008A377A
+bool ExperienceTracker::Gain_Exp_For_Level(int32_t levels_to_gain, bool apply_multiplier)
+{
+    const auto new_level = std::min<VeterancyLevel>(m_currentLevel + levels_to_gain, VeterancyLevel::VETERANCY_COUNT - 1);
+
+    if (new_level <= m_currentLevel) {
+        return false;
+    }
+
+    const auto exp_for_level_gain = m_parent->Get_Template()->Get_Experience_Required(new_level) - m_currentExperience;
+
+    Add_Experience_Points(exp_for_level_gain, apply_multiplier);
+    return true;
+}
+
+// zh: 0x0061B8F0 wb: 0x008A37E4
+bool ExperienceTracker::Can_Gain_Exp_For_Level(int32_t levels_to_gain) const
+{
+    return m_currentLevel != std::min<VeterancyLevel>(m_currentLevel + levels_to_gain, VeterancyLevel::VETERANCY_COUNT - 1);
+}
+
+// zh: 0x0061B910 wb: 0x008A381C
+void ExperienceTracker::Add_Experience_Points(int32_t experience_gain, bool apply_multiplier)
+{
+    if (m_experienceSink != ObjectID::OBJECT_UNK) {
+        auto *sink_object = g_theGameLogic->Find_Object_By_ID(m_experienceSink);
+        if (sink_object != nullptr) {
+            sink_object->Get_Experience_Tracker()->Add_Experience_Points(
+                experience_gain * m_experienceMultiplier, apply_multiplier);
+            return;
+        }
+    }
+
+    if (!Is_Trainable()) {
+        return;
+    }
+
+    const auto old_level = m_currentLevel;
+    if (apply_multiplier) {
+        experience_gain *= m_experienceMultiplier;
```

Comment:
experience_gain is typecast to float before the multiply happens and then back to int after the multiply. Should those casts be there?

Reply:
not required it will do a promotion of the int to a float, do the multiply, then demote back to int.